File: OSCommon/org/optimizationservices/oscommon/representationparser/OSbLReader.cs
using System;
using System.Collections;
using System.Xml;

using org.optimizationservices.oscommon.util;
using org.optimizationservices.oscommon.localinterface;
using org.optimizationservices.oscommon.datastructure.osbenchmark;

namespace org.optimizationservices.oscommon.representationparser{
	///	<summary>
	///	The	<c>OSbLReader</c> class	parses an OSbL instance	into a DOM tree	and
	///	provides a set of "get"	methods	that can be	
	///	used to	retrieve different pieces of information of	the	instance.
	///	@author	Jun	Ma	
	///	@version 1.0, 09/01/2005
	///	@since OS 1.0
	///	@copyright (c) 2005
	///	</summary>
	public class OSbLReader	: OSgLReader{

		/// <summary>
		/// m_osBenchmark holds the standard OSBenchmark, which is a local interface for 
		/// storing Optimization Services benchmark result. 
		/// </summary>
		protected OSBenchmark m_osBenchmark = null;

		///	<summary>
		///	m_otherBenchmarkHashMap holds a hash map of other alalyses.
		///	</summary>
		protected Hashtable	m_otherBenchmarkHashMap = null;

		///	<summary>
		///	m_otherBenchmarkDescriptionHashMap holds a hash map of other benchmarks' descriptions.
		///	</summary>
		protected Hashtable	m_otherBenchmarkDescriptionHashMap = null;

		///	<summary>
		///	constructor.
		///	</summary>
		///	<param name="validate">holds whether the reader	should be validating against the schema	or not.</param>
		public OSbLReader(bool validate){
			m_bXsdValidate = validate;
		}//constructor
	
		///	<summary>
		///	Default	constructor. 
		///	</summary>
		public OSbLReader(){
		}//constructor

		/// <summary>
		/// get the standard OSBenchmark, a local interface for 
		/// storing Optimization Services benchmark result.
		/// </summary>
		/// <returns>the OSBenchmark. </returns>
		public OSBenchmark getOSBenchmark(){
			if(m_osBenchmark != null){
				return m_osBenchmark;
			}
			m_osBenchmark = new OSBenchmark();		
			if(!m_osBenchmark.setServiceURI(getServiceURI())) throw new Exception("setServiceURI Unsuccessful");
			if(!m_osBenchmark.setServiceName(getServiceName())) throw new Exception("setServiceName Unsuccessful");
			if(!m_osBenchmark.setBenchmarkTime(getBenchmarkTime())) throw new Exception("setBenchmarkTime Unsuccessful");
			if(!m_osBenchmark.setBenchmarkMessage(getBenchmarkMessage())) throw new Exception("setBenchmarkMessage Unsuccessful");
			if(!m_osBenchmark.setBenchmarkScores(getBenchmarkScores())) throw new Exception("setBenchmarkScores Unsuccessful");
			if(!m_osBenchmark.setBenchmarkStatistics(getBenchmarkStatistics())) throw new Exception("setBenchmarkStatistics Unsuccessful");
			if(!m_osBenchmark.setOptimizationBenchmark(getOptimizationBenchmark())) throw new Exception("setOptimizationBenchmark Unsuccessful");
			if(!m_osBenchmark.setOtherBenchmarks(getOtherBenchmarkNames(), getOtherBenchmarkDescriptions(), getOtherBenchmarkValues())) throw new Exception("setOtherBenchmarks Unsuccessful");
			return m_osBenchmark;
		}//getOSBenchmark

		/// <summary>
		/// get the service uri
		/// </summary>
		/// <returns>the service uri</returns>
		public string getServiceURI(){
			XmlElement eBenchmarkHeader = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
			string sServiceURI = XMLUtil.getElementValueByName(eBenchmarkHeader, "serviceURI");
			if(sServiceURI == null) sServiceURI = "";
			return sServiceURI;
		}//getServiceURI

		/// <summary>
		/// get the service name
		/// </summary>
		/// <returns>the service name</returns>
		public string getServiceName(){
			XmlElement eBenchmarkHeader = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
			string sServiceName = XMLUtil.getElementValueByName(eBenchmarkHeader, "serviceName");
			if(sServiceName == null) sServiceName = "";
			return sServiceName;
		}//getServiceName

			
		/// <summary>
		/// Get time of the last benchmark.
		/// </summary>
		/// <returns>the time of the last benchmark.</returns>
		public DateTime getBenchmarkTime(){
			XmlElement eBenchmarkHeader = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
			string sTime = XMLUtil.getElementValueByName(eBenchmarkHeader, "time");
			if(sTime == null) return new DateTime(1970, 1, 1, 0, 0, 0);
			DateTime time = XMLUtil.createNativeDateTime(sTime);
			return time;
		}//getBenchmarkTime

		/// <summary>
		/// Get the benchmark message. 
		/// </summary>
		/// <returns>the benchmark message. </returns>
		public string getBenchmarkMessage(){
			XmlElement eBenchmarkHeader = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
			string sMessage = XMLUtil.getElementValueByName(eBenchmarkHeader, "message");
			if(sMessage == null) sMessage = "";
			return sMessage;	
		}//getBenchmarkMessage

		
		/// <summary>
		/// Get the benchmark scores. 
		/// </summary>
		/// <returns>the benchmark scores. </returns>
		public BenchmarkScores getBenchmarkScores(){
			XmlElement eBenchmarkData = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
			XmlElement eScores = (XmlElement)XMLUtil.findChildNode(eBenchmarkData, "scores");
			if(eScores == null) return null;
			BenchmarkScores benchmarkScores = new BenchmarkScores();
			string sValue;
			sValue = XMLUtil.getElementValueByName(eScores, "overall");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.overall = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "expertAssessment");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.expertAssessment = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "userAssessment");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.userAssessment = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "reputation");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.reputation = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "popularity");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.popularity = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "owner");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.owner = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "support");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.support = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "system");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.system = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "service");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.service = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eScores, "computation");
			if(sValue != null && sValue.Length > 0){
				benchmarkScores.computation = Double.Parse(sValue);			
			}
			return benchmarkScores;
		}//getBenchmarkScores

		/// <summary>
		/// Get the benchmark statistics.
		/// </summary>
		/// <returns>the benchmark statistics.</returns>
		public BenchmarkStatistics getBenchmarkStatistics(){
			XmlElement eBenchmarkData = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
			XmlElement eStatistics = (XmlElement)XMLUtil.findChildNode(eBenchmarkData, "statistics");
			if(eStatistics == null) return null;
			BenchmarkStatistics benchmarkStatistics = new BenchmarkStatistics();
			string sValue;
			sValue = XMLUtil.getElementValueByName(eStatistics, "averageJobs");
			if(sValue != null && sValue.Length > 0){
				benchmarkStatistics.averageJobs = Double.Parse(sValue);			
			}
			sValue = XMLUtil.getElementValueByName(eStatistics, "averageWaitTime");
			if(sValue != null && sValue.Length > 0){
				benchmarkStatistics.averageWaitTime = Double.Parse(sValue);			
			}		
			sValue = XMLUtil.getElementValueByName(eStatistics, "averageServiceTime");
			if(sValue != null && sValue.Length > 0){
				benchmarkStatistics.averageServiceTime = Double.Parse(sValue);			
			}					
			return benchmarkStatistics;
		}//getBenchmarkStatistics

		/// <summary>
		/// Get the optimization benchmark information. 
		/// </summary>
		/// <returns>the optimization benchmark information. </returns>
		public OptimizationBenchmark getOptimizationBenchmark(){
			XmlElement eBenchmarkData = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
			XmlElement eOptimization = (XmlElement)XMLUtil.findChildNode(eBenchmarkData, "optimization");
			if(eOptimization == null) return null;
			OptimizationBenchmark optimizationBenchmark = new OptimizationBenchmark();
			return optimizationBenchmark;
		}//getOptimizationBenchmark

		/// <summary>
		/// Get the hash map of other benchmarks. 
		/// </summary>
		/// <returns>the hash map of other benchmarks.</returns>
		public Hashtable getOtherBenchmarks(){
			if(m_otherBenchmarkHashMap != null) return m_otherBenchmarkHashMap;
			m_otherBenchmarkHashMap = new Hashtable();
			m_otherBenchmarkDescriptionHashMap = new Hashtable();

			XmlElement eBenchmarkData = (XmlElement)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
			ArrayList vOtherElement = XMLUtil.getChildElementsByTagName(eBenchmarkData, "other");
			int iNls	= vOtherElement==null?0:vOtherElement.Count;
			for(int i = 0; i < iNls; i++){
				XmlElement eOther = (XmlElement)(vOtherElement[i]);
				XmlNamedNodeMap	attributes =  eOther.Attributes;
				int n =attributes.Count;
				string sName = "";
				string sDescription = "";
				string sValue = XMLUtil.getElementValue(eOther);
				for (int j = 0; j < n; j++){
					XmlNode	attr = attributes.Item(j);
					string sAttributeName  = attr.LocalName;
					string sAttributeValue = attr.Value;
					if (sAttributeName.Equals("name")){
						sName = sAttributeValue;
					}
					else if (sAttributeName.Equals("description")){
						sDescription = sAttributeValue;
					}
				}
				m_otherBenchmarkHashMap.Add(sName, sValue);
				m_otherBenchmarkDescriptionHashMap.Add(sName, sDescription);
			}
			return m_otherBenchmarkHashMap;
		}//getOtherBenchmarks

		/// <summary>
		/// Get the number of other benchmarks.  
		/// </summary>
		/// <returns>the number of other benchmarks. </returns>
		public int getOtherBenchmarkNumber(){
			getOtherBenchmarks();
			return m_otherBenchmarkHashMap.Count;
		}//getOtherBenchmarkNumber

		/// <summary>
		/// Get the string value from the other benchmark hash map. 
		/// </summary>
		/// <param name="name">holds the name of benchmark to get. </param>
		/// <returns>string value from the other benchmark hash map, null if none. </returns>
		public string getOtherBenchmarkValueByName(string name){
			getOtherBenchmarks();
			if(m_otherBenchmarkHashMap.Contains(name)){
				return (string)m_otherBenchmarkHashMap[name];
			}
			else{ 
				return null;
			}
		}//getOtherBenchmarkValueByName

		/// <summary>
		/// Get the string value from the other benchmark hash map. 
		/// </summary>
		/// <param name="name">holds the name of benchmark to get. </param>
		/// <returns>string value from the other benchmark hash map, null if none. </returns>
		public string getOtherBenchmarkDescriptionByName(string name){
			getOtherBenchmarks();
			if(m_otherBenchmarkDescriptionHashMap.Contains(name)){
				return (string)m_otherBenchmarkDescriptionHashMap[name];
			}
			else{ 
				return null;
			}
		}//getOtherBenchmarkDescriptionByName

		/// <summary>
		/// Get the names of all other benchmarks. 
		/// </summary>
		/// <returns>the names of all other benchmarks. </returns>
		public string[] getOtherBenchmarkNames(){
			getOtherBenchmarks();	
			System.Collections.ICollection nameCollection = m_otherBenchmarkHashMap.Keys;
			string[] msName = new string[nameCollection.Count];
			IDictionaryEnumerator dictionaryEnumerator = m_otherBenchmarkHashMap.GetEnumerator();
			int i = 0;
			while(dictionaryEnumerator.MoveNext()){
				msName[i] = (string)dictionaryEnumerator.Key;
				i++;
			}
			return msName;
		}//getOtherBenchmarkNames

		/// <summary>
		/// Get the values of all other benchmarks. 
		/// </summary>
		/// <returns>the values of all other benchmarks. </returns>
		public string[] getOtherBenchmarkValues(){
			getOtherBenchmarks();	
			System.Collections.ICollection valueCollection = m_otherBenchmarkHashMap.Values;
			string[] msValue = new string[valueCollection.Count];
			IDictionaryEnumerator dictionaryEnumerator = m_otherBenchmarkHashMap.GetEnumerator();
			int i = 0;
			while(dictionaryEnumerator.MoveNext()){
				msValue[i] = (string)dictionaryEnumerator.Value;
				i++;
			}
			return msValue;
		}//getOtherBenchmarkValues

		/// <summary>
		/// Get the descriptions of all other benchmarks. 
		/// </summary>
		/// <returns>the descriptions of all other benchmarks. </returns>
		public string[] getOtherBenchmarkDescriptions(){
			getOtherBenchmarks();	
			System.Collections.ICollection descriptionCollection = m_otherBenchmarkDescriptionHashMap.Values;
			string[] msDescription = new string[descriptionCollection.Count];
			IDictionaryEnumerator dictionaryEnumerator = m_otherBenchmarkDescriptionHashMap.GetEnumerator();
			int i = 0;
			while(dictionaryEnumerator.MoveNext()){
				msDescription[i] = (string)dictionaryEnumerator.Value;
				i++;
			}
			return msDescription;
		}//getOtherBenchmarkDescriptions

	}//class OSbLReader
}//namespace

File: OSjava/OSCommon/src/org/optimizationservices/oscommon/representationparser/OSbLReader.java
/**
 * @(#)OSbLReader 1.0 03/14/2004
 *
 * Copyright (c) 2004
 */
package org.optimizationservices.oscommon.representationparser;

import java.util.Collection;
import java.util.GregorianCalendar;
import java.util.HashMap;
import java.util.Set;
import java.util.Vector;

import org.optimizationservices.oscommon.datastructure.osbenchmark.BenchmarkScores;
import org.optimizationservices.oscommon.datastructure.osbenchmark.BenchmarkStatistics;
import org.optimizationservices.oscommon.datastructure.osbenchmark.OptimizationBenchmark;
import org.optimizationservices.oscommon.localinterface.OSBenchmark;
import org.optimizationservices.oscommon.util.IOUtil;
import org.optimizationservices.oscommon.util.OSParameter;
import org.optimizationservices.oscommon.util.XMLUtil;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;

/**
 * The <c>OSbLReader</c> class parses an OSbL instance into a DOM tree and
 * provides a set of "get" methods that can be used to retrieve different 
 * pieces of information of the instance.
 * </p>
 *
 * @author Robert Fourer, Jun Ma, Kipp Martin
 * @version 1.0, 03/14/2004
 * @since OS 1.0
 */
public class OSbLReader extends OSgLReader{
	/**
	 * m_osBenchmark holds the standard OSBenchmark, which is a local interface for 
	 * storing Optimization Services benchmark result. 
	 */
	protected OSBenchmark m_osBenchmark = null;
	
	/**
	 * m_otherBenchmarkHashMap holds a hash map of other alalyses.
	 */
	protected HashMap <String, String>	m_otherBenchmarkHashMap = null;

	/**
	 * m_otherBenchmarkDescriptionHashMap holds a hash map of other benchmarks' descriptions.
	 */
	protected HashMap <String, String>	m_otherBenchmarkDescriptionHashMap = null;

	/**
	 * constructor.
	 * @param validate holds whether the reader	should be validating against the schema	or not.
	 */
	public OSbLReader(boolean validate){
		m_bXsdValidate = validate;
	}//constructor

	/**
	 * Default	constructor. 
	 *
	 */
	public OSbLReader(){
	}//constructor
	
	/**
	 * get the standard OSBenchmark, a local interface for 
	 * storing Optimization Services benchmark result.
	 * 
	 * @return the OSBenchmark. 
	 * @throws Exception if there are errors getting the OSBenchmark. 
	 */
   	public OSBenchmark getOSBenchmark() throws Exception{
		if(m_osBenchmark != null){
			return m_osBenchmark;
		}
		m_osBenchmark = new OSBenchmark();		
		if(!m_osBenchmark.setServiceURI(getServiceURI())) throw new Exception("setServiceURI Unsuccessful");
		if(!m_osBenchmark.setServiceName(getServiceName())) throw new Exception("setServiceName Unsuccessful");
		if(!m_osBenchmark.setBenchmarkTime(getBenchmarkTime())) throw new Exception("setBenchmarkTime Unsuccessful");
		if(!m_osBenchmark.setBenchmarkMessage(getBenchmarkMessage())) throw new Exception("setBenchmarkMessage Unsuccessful");
		if(!m_osBenchmark.setBenchmarkScores(getBenchmarkScores())) throw new Exception("setBenchmarkScores Unsuccessful");
		if(!m_osBenchmark.setBenchmarkStatistics(getBenchmarkStatistics())) throw new Exception("setBenchmarkStatistics Unsuccessful");
		if(!m_osBenchmark.setOptimizationBenchmark(getOptimizationBenchmark())) throw new Exception("setOptimizationBenchmark Unsuccessful");
		if(!m_osBenchmark.setOtherBenchmarks(getOtherBenchmarkNames(), getOtherBenchmarkDescriptions(), getOtherBenchmarkValues())) throw new Exception("setOtherBenchmarks Unsuccessful");
		return m_osBenchmark;
   	}//getOSBenchmark
   	
	/**
	 * Get the service uri.
	 * @return the service uri
	 */
	public String getServiceURI(){
		Element eBenchmarkHeader = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
		String sServiceURI = XMLUtil.getElementValueByName(eBenchmarkHeader, "serviceURI");
		if(sServiceURI == null) sServiceURI = "";
		return sServiceURI;	
	}//getServiceURI
	
	/**
	 * Get the service name.
	 * 
	 * @return the service name
	 */
	public String getServiceName(){
		Element eBenchmarkHeader = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
		String sServiceName = XMLUtil.getElementValueByName(eBenchmarkHeader, "serviceName");
		if(sServiceName == null) sServiceName = "";
		return sServiceName;
	}//getServiceName
	
  	/**
	 * Get time of the last benchmark.
	 * 
	 * @return the time of the last benchmark.
	 */
	public GregorianCalendar getBenchmarkTime(){
		Element eBenchmarkHeader = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
		String sTime = XMLUtil.getElementValueByName(eBenchmarkHeader, "time");
		if(sTime == null) return null;
		GregorianCalendar time = XMLUtil.createNativeDateTime(sTime);
		return time;	
	}//getBenchmarkTime
	
   	/**
	 * Get the benchmark message. 
	 * 
	 * @return the benchmark message. 
	 */
	public String getBenchmarkMessage(){
		Element eBenchmarkHeader = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkHeader");
		String sMessage = XMLUtil.getElementValueByName(eBenchmarkHeader, "message");
		if(sMessage == null) sMessage = "";
		return sMessage;
	}//getBenchmarkMessage
	
	/**
	 * Get the benchmark scores. 
	 * 
	 * @return the benchmark scores. 
	 */
	public BenchmarkScores getBenchmarkScores(){
		Element eBenchmarkData = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
		Element eScores = (Element)XMLUtil.findChildNode(eBenchmarkData, "scores");
		if(eScores == null) return null;
		BenchmarkScores benchmarkScores = new BenchmarkScores();
		String sValue;
		sValue = XMLUtil.getElementValueByName(eScores, "overall");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.overall = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "expertAssessment");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.expertAssessment = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "userAssessment");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.userAssessment = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "reputation");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.reputation = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "popularity");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.popularity = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "owner");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.owner = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "support");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.support = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "system");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.system = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "service");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.service = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eScores, "computation");
		if(sValue != null && sValue.length() > 0){
			benchmarkScores.computation = Double.parseDouble(sValue);			
		}
		return benchmarkScores;
	}//getBenchmarkScores
	
	/**
	 * Get the benchmark statistics. 
	 * 
	 * @return the benchmark statistics. 
	 */
	public BenchmarkStatistics getBenchmarkStatistics(){
		Element eBenchmarkData = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
		Element eStatistics = (Element)XMLUtil.findChildNode(eBenchmarkData, "statistics");
		if(eStatistics == null) return null;
		BenchmarkStatistics benchmarkStatistics = new BenchmarkStatistics();
		String sValue;
		sValue = XMLUtil.getElementValueByName(eStatistics, "averageJobs");
		if(sValue != null && sValue.length() > 0){
			benchmarkStatistics.averageJobs = Double.parseDouble(sValue);			
		}
		sValue = XMLUtil.getElementValueByName(eStatistics, "averageWaitTime");
		if(sValue != null && sValue.length() > 0){
			benchmarkStatistics.averageWaitTime = Double.parseDouble(sValue);			
		}		
		sValue = XMLUtil.getElementValueByName(eStatistics, "averageServiceTime");
		if(sValue != null && sValue.length() > 0){
			benchmarkStatistics.averageServiceTime = Double.parseDouble(sValue);			
		}		
		return benchmarkStatistics;
	}//getBenchmarkStatistics
	
	/**
	 * Get the optimization benchmark information. 
	 * 
	 * @return the optimization benchmark information. 
	 */
	public OptimizationBenchmark getOptimizationBenchmark(){
		Element eBenchmarkData = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
		Element eOptimization = (Element)XMLUtil.findChildNode(eBenchmarkData, "optimization");
		if(eOptimization == null) return null;
		OptimizationBenchmark optimizationBenchmark = new OptimizationBenchmark();
		return optimizationBenchmark;
	}//getOptimizationBenchmark
	
	/**
	 * Get the hash map of other benchmarks. 
	 * 
	 * @return the hash map of other benchmarks.
	 */
	public HashMap <String, String> getOtherBenchmarks(){
		if(m_otherBenchmarkHashMap != null) return m_otherBenchmarkHashMap;
		m_otherBenchmarkHashMap = new HashMap <String, String>();
		m_otherBenchmarkDescriptionHashMap = new HashMap <String, String>();
		Element eBenchmarkData = (Element)XMLUtil.findChildNode(m_eRoot, "benchmarkData");
		Vector<Element> vOtherElement = XMLUtil.getChildElementsByTagName(eBenchmarkData, "other");
		int iNls = vOtherElement==null?0:vOtherElement.size();
		for(int i = 0; i < iNls; i++){
			Element eOther = (Element)(vOtherElement.elementAt(i));
			NamedNodeMap	attributes =  eOther.getAttributes();
			int n =attributes.getLength();
			String sName = "";
			String sDescription = "";
			String sValue = XMLUtil.getElementValue(eOther);
			for (int j = 0; j < n; j++){
				Node	attr = attributes.item(j);
				String sAttributeName  = attr.getNodeName();
				String sAttributeValue = attr.getNodeValue();
				if (sAttributeName.equals("name")){
					sName = sAttributeValue;
				}
				else if (sAttributeName.equals("description")){
					sDescription = sAttributeValue;
				}
			}
			m_otherBenchmarkHashMap.put(sName, sValue);
			m_otherBenchmarkDescriptionHashMap.put(sName, sDescription);
		}
		return m_otherBenchmarkHashMap;
	}//getOtherBenchmarks
		
	/**
	 * Get the number of other benchmarks.  
	 * 
	 * @return the number of other benchmarks. 
	 */
	public int getOtherBenchmarkNumber(){
		getOtherBenchmarks();
		return m_otherBenchmarkHashMap.size();
	}//getOtherBenchmarkNumber

	/**
	 * Get the String value from the other benchmark hash map. 
	 * 
	 * @param name holds the name of benchmark to get. 
	 * @return String value from the other benchmark hash map, null if none.
	 */
	public String getOtherBenchmarkValueByName(String name){
		getOtherBenchmarks();
		if(m_otherBenchmarkHashMap.containsKey(name)){
			return (String)m_otherBenchmarkHashMap.get(name);
		}
		else{ 
			return null;
		}
	}//getOtherBenchmarkValueByName

	/**
	 * Get the String value from the other benchmark hash map. 
	 * 
	 * @param name holds the name of benchmark to get.
	 * @return String value from the other benchmark hash map, null if none.
	 */
	public String getOtherBenchmarkDescriptionByName(String name){
		getOtherBenchmarks();
		if(m_otherBenchmarkDescriptionHashMap.containsKey(name)){
			return (String)m_otherBenchmarkDescriptionHashMap.get(name);
		}
		else{ 
			return null;
		}
	}//getOtherBenchmarkDescriptionByName

	
	/**
	 * Get the names of all other benchmarks. 
	 * 
	 * @return the names of all other benchmarks.
	 */
	public String[] getOtherBenchmarkNames(){
		getOtherBenchmarks();	
		Set nameSet = m_otherBenchmarkHashMap.keySet();
		Object[] moName = nameSet.toArray();
		String[] msName = new String[nameSet.size()];
		for (int i = 0; i < nameSet.size(); i++) {
			msName[i] = moName[i].toString();
		}
		return msName;		
	}//getOtherBenchmarkNames
	
	/**
	 *  Get the values of all other benchmarks. 
	 * 
	 * @return the values of all other benchmarks. 
	 */
	public String[] getOtherBenchmarkValues(){
		getOtherBenchmarks();		
		Collection valueCollection = m_otherBenchmarkHashMap.values();
		Object[] moValue = valueCollection.toArray();
		String[] msValue = new String[valueCollection.size()];
		for (int i = 0; i < valueCollection.size(); i++) {
			msValue[i] = moValue[i].toString();
		}
		return msValue;		
	}//getOtherBenchmarkValues

	
	/**
	 * Get the descriptions of all other benchmarks. 
	 * 
	 * @return the descriptions of all other benchmarks.
	 */
	public String[] getOtherBenchmarkDescriptions(){
		getOtherBenchmarks();			
		Collection valueCollection = m_otherBenchmarkDescriptionHashMap.values();
		Object[] moValue = valueCollection.toArray();
		String[] msDescription = new String[valueCollection.size()];
		for (int i = 0; i < valueCollection.size(); i++) {
			msDescription[i] = moValue[i].toString();
		}
		return msDescription;		
	}//getOtherBenchmarkDescriptions
	
	
	/**
	 * main for test purposes.
	 *
	 * </p>
	 *
	 * @param argv command line arguments.
	 */
	public static void main(String[] argv){
		OSbLReader osblReader = new OSbLReader(false);
		System.out.println(IOUtil.readStringFromFile(OSParameter.CODE_HOME + "OSRepository/test/osbl/osbl.osbl"));
		System.out.println(osblReader.readFile(OSParameter.CODE_HOME + "OSRepository/test/osbl/osbl.osbl"));
		System.out.println(osblReader.getBenchmarkMessage());
		System.out.println(XMLUtil.createXSDateTime(osblReader.getBenchmarkTime()));
		System.out.println(osblReader.getServiceName());
		System.out.println(osblReader.getServiceURI());
		BenchmarkScores benchmarkScores = osblReader.getBenchmarkScores();
		if(benchmarkScores != null){
			System.out.println(benchmarkScores.computation);
			System.out.println(benchmarkScores.expertAssessment);
			System.out.println(benchmarkScores.overall);
			System.out.println(benchmarkScores.owner);
			System.out.println(benchmarkScores.popularity);
			System.out.println(benchmarkScores.reputation);
			System.out.println(benchmarkScores.service);
			System.out.println(benchmarkScores.support);
			System.out.println(benchmarkScores.system);
			System.out.println(benchmarkScores.userAssessment);
		}
		BenchmarkStatistics benchmarkStatistics = osblReader.getBenchmarkStatistics();
		if(benchmarkStatistics != null){
			System.out.println(benchmarkStatistics.averageJobs);
			System.out.println(benchmarkStatistics.averageServiceTime);
			System.out.println(benchmarkStatistics.averageWaitTime);
		}
		osblReader.getOptimizationBenchmark();
		System.out.println(osblReader.getOtherBenchmarkNumber());
		String sValue = osblReader.getOtherBenchmarkValueByName("ad");
		if(sValue == null) System.out.println("null");
		else System.out.println(sValue);
		String sDescription = osblReader.getOtherBenchmarkDescriptionByName("a");
		System.out.println(sDescription);
		for(int i=0;i<osblReader.getOtherBenchmarkNumber();i++){
			System.out.println(osblReader.getOtherBenchmarkNames()[i]);
			System.out.println(osblReader.getOtherBenchmarkValues()[i]);
			System.out.println(osblReader.getOtherBenchmarkDescriptions()[i]);
		}
		
		osblReader.writeToStandardOutput();
	}//main
	
}//class OSbLReader
